src/game/common/thing/thingfactory.cpp

```diff
@@ -13,15 +13,167 @@
+
+    if (!m_nextTemplateID) {
+        captainslog_debug("m_nextTemplateID wrapped to zero");
+    }
+
+    new_template->Friend_Set_Template_Name(name);
+    Add_Template(new_template);
+    return new_template;
+}
+
+ThingTemplate *ThingFactory::New_Override(ThingTemplate *thing_template)
+{
+    captainslog_dbgassert(thing_template, "NULL 'parent' thing template");
+    captainslog_dbgassert(Find_Template(thing_template->Get_Name(), true),
+        "Thing template '%s' not in master list",
+        thing_template->Get_Name());
+
+    ThingTemplate *new_template = NEW_POOL_OBJ(ThingTemplate);
+    ThingTemplate *overrid = (ThingTemplate *)thing_template->Get_Final_Override();
+    *new_template = *overrid;
+    new_template->Set_Copied_From_Default();
+    new_template->Set_Is_Allocated();
+    overrid->Set_Next(new_template);
+    return new_template;
+}
+
+void ThingFactory::Free_Database()
+{
+    while (m_firstTemplate) {
+        ThingTemplate *t = m_firstTemplate;
+        m_firstTemplate = t->Friend_Get_Next_Template();
+        t->Delete_Instance();
+    }
+
+    m_templateMap.clear();
+}
+
+void ThingFactory::PostProcessLoad()
+{
+    for (ThingTemplate *i = m_firstTemplate; i; i = i->Friend_Get_Next_Template()) {
+        i->Resolve_Names();
+    }
+}
+
+void ThingFactory::Reset()
+{
+    ThingTemplate *t = m_firstTemplate;
+
+    while (t) {
+        bool first = false;
+        ThingTemplate *t2 = t->Friend_Get_Next_Template();
+
+        if (t == m_firstTemplate) {
+            first = true;
+        }
+
+        Utf8String str(t->Get_Name());
+        Overridable *o = t->Delete_Overrides();
+
+        if (!o) {
+            if (first) {
+                m_firstTemplate = t2;
+            }
+        }
+
+        if (!o) {
+            m_templateMap.erase(str);
+        }
+
+        t = t2;
+    }
+}
+
+void ThingFactory::Add_Template(ThingTemplate *tmplate)
+{
+    if (m_templateMap.find(tmplate->Get_Name()) != m_templateMap.end()) {
+        captainslog_debug("Duplicate Thing Template name found: %s", tmplate->Get_Name());
+    }
+
+    tmplate->Friend_Set_Next_Template(m_firstTemplate);
+    m_firstTemplate = tmplate;
+    m_templateMap[tmplate->Get_Name()] = tmplate;
+}
+
+const ThingTemplate *ThingFactory::Find_Template_By_ID(unsigned short id)
+{
+    for (ThingTemplate *i = m_firstTemplate; i; i = i->Friend_Get_Next_Template()) {
+        if (i->Get_Template_ID() == id) {
+            return i;
+        }
+    }
+
+    captainslog_debug("template %d not found", id);
+    return nullptr;
+}
+
+const ThingTemplate *ThingFactory::Find_Template_Internal(const Utf8String &name, bool b)
+{
+    auto i = m_templateMap.find(name);
+
+    if (i != m_templateMap.end()) {
+        return i->second;
+    }
+
+    if (!strncmp(name, "***TESTING", strlen("***TESTING"))) {
+        ThingTemplate *tmplate = New_Template("Un-namedTemplate");
+        tmplate->Init_For_LTA(name);
+        m_templateMap.erase("Un-namedTemplate");

```

Comment:
isn't this a memory leak? Erasing the pointer which we haven't destructed? Perhaps add a TODO to investigate.

Reply:
Its not a memory leak.
The erase call is removing the pointer stored in tmplate (which is then added again under the new name)